src/main/scala/chisel3/util/experimental/ForceNames.scala

```diff
@@ -0,0 +1,290 @@
+    renames.get(target) match {
+      case None => List(this)
+      case Some(newTargets) if newTargets.size > 1 =>
+        throw CustomTransformException(
+          new FirrtlUserException(
+            s"Cannot force the name of $target to $name because it is renamed to $newTargets." +
+              " Perhaps $target is not a ground type?"
+          )
+        )
+      case Some(newTargets) => newTargets.map(t => duplicate(t))
+    }
+  }
+}
+
+/** Contains utility functions for [[ForceNamesTransform]] */
+object ForceNamesTransform {
+  /** Returns the [[IsModule]] which is referred to, or if a [[ReferenceTarget]], the enclosing [[IsModule]]
+    * @param a signal/instance/module
+    * @return referring IsModule
+    */
+  def referringIsModule(a: IsMember): IsModule = a match {
+    case b: ModuleTarget    => b
+    case b: InstanceTarget  => b.targetParent
+    case b: ReferenceTarget => b.copy(component = Nil).targetParent.asInstanceOf[IsModule]
+  }
+
+  /** Returns a function which returns all instance paths to a given IsModule
+    * @param graph
+    * @return
+    */
+  def allInstancePaths(graph: InstanceKeyGraph): IsModule => List[List[(Instance, OfModule)]] = {
+    val lookup: String => List[List[(Instance, OfModule)]] =
+      str => graph.findInstancesInHierarchy(str)
+        .view
+        .map(_.map(_.toTokens).toList)
+        .toList
+    allInstancePaths(lookup) _
+  }
+
+  /** Returns a function which returns all instance paths to a given IsModule
+    * Uses Lists because we want to recurse on inner List (head :: tail)-style
+    * @param lookup given a module, return all paths to the module
+    * @param target target to get all instance paths to
+    * @return
+    */
+  def allInstancePaths(lookup: String => List[List[(Instance, OfModule)]])
+                      (target: IsModule): List[List[(Instance, OfModule)]] = {
+    target match {
+      case ModuleTarget(circuit, module) =>
+        if (circuit == module) List(List((Instance(module), OfModule(module))))
+        else lookup(module)
+      case it: InstanceTarget =>
+        val itPath = it.asPath.toList
+        if (it.module == it.circuit) List((Instance(it.module), OfModule(it.module)) :: itPath)
+        else lookup(it.module).map(_ ++ itPath)
+    }
+  }
+
+
+  /** Builds the map of module name to map of old signal/instance name to new signal/instance name
+    *
+    * @param state CircuitState to operate on
+    * @param igraph built instance key graph from state's circuit
+    * @return
+    */
+  def buildForceNameMap(state: CircuitState,
+                     igraph: => InstanceKeyGraph
+                    ): Option[Map[String, Map[String, String]]] = {
+    val forceNames = state.annotations.collect { case f: ForceNameAnnotation => f }
+    val badNames = mutable.HashSet[ForceNameAnnotation]()
+    val allNameMaps = forceNames.groupBy { case f => referringIsModule(f.target) }.mapValues { value =>
```

Comment:
`mapValues` returns a `MapView` which means it will rerun this closure each time you access each value. Since you're not even using `allNameMaps` as a `Map` below, I'd suggest casting to a `Seq` or something, but at the very least you need to force to an eager collection.

Reply:
So I casted it to a Seq, and then back to a Map because its the return value. Do you have a better suggestion?

Reply:
That's fine, or you can instead just to `.toMap` instead of `.toSeq` and keep it as a `Map` the whole time, I didn't realize it was the return value.